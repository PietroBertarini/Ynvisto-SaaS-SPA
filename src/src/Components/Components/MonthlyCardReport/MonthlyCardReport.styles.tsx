import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

const styledBy = (property, mapping) => (props) => mapping[props[property]];
export const useStyles = makeStyles((theme: Theme) => createStyles({
  titleContainer: {
    width: '100%',
    height: '4vw',
    paddingTop: '0!important',
    paddingBottom: '0vw!important',
    paddingLeft: '0!important',
    paddingRight: '0!important',
    background: '#28282F',
  },
  card: {
    width: '100%',
    height: '100%',
    borderRadius: 20,
    background: 'transparent',
  },
  content: {
    width: '100%',
    height: '20vw',
    minHeight: 300,
    background: '#1E1E23',
  },
  title: {
    color: '#FFFFFF',
    fontFamily: 'Zurich',
    fontWeight: 'bold',
  },
  actionArea: {
    display: 'inline-block',
  },
  titleContainerBox: {
    width: '100%',
    height: '100%',
  },
  headerMonetaryValue: {
    textAlign: 'center',
    color: styledBy('status', {
      GOOD: '#05E985',
      WARNING: '#E9C46A',
      BAD: '#EB5050',
    }),
    marginBottom: '1vw',
    fontWeight: 'bold',
  },
  modal: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: '100vw',
    '& .MuiDialog-container': {
      width: '100%',
    },
  },
  paper: {
    backgroundColor: '#1E1E23',
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
    width: '25vw',
  },
  showDiv: {
    width: '90%',
    color: '#ccc',
    borderTop: '1px solid #ccc',
    padding: '20',
    position: 'absolute',
    bottom: '0',
  },
  showMore: {
    padding: '10',
    textAlign: 'center',
  },
}));
