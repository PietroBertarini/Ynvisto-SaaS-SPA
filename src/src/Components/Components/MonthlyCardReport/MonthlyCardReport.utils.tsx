import '../../../Theme/sweetalert2/sweetalert2.scss';
import { allTabs } from '../../Pages/TaxReport/TaxReport.utils';
import MonthlyTaxReport from '../../../Classes/MonthlyTaxReport/MonthlyTaxReport.class';
import { ITaxReport } from '../../../Redux/Services/TaxReports/TaxReports.actions';

export interface ITaxdDetailInfo {
  commonOps?: {
    totalSell: number,
    monthProfit: number,
    taxableProfit: number,
    compensatedLoss: number,
    calcBase: number,
    taxDue: number,
  },
  dayTrade?: {
    taxableProfit: number,
    compensatedLoss: number,
    calcBase: number,
    taxDue: number,
  },
  fii?: {
    liquidResult: number,
    compensatedLoss: number,
    calcBase: number,
    taxDue: number,
  },
  finalResults: {
    commonOpTax: number,
    daytradeTax: number,
    withholdingTax: number,
    totalTax: number,
    fiiTax: number,
    previousDarf: number,
    payableTax: number,
  },
}

const defaultTaxdDetailInfo: ITaxdDetailInfo = {
  finalResults: {
    commonOpTax: 0,
    daytradeTax: 0,
    withholdingTax: 0,
    totalTax: 0,
    fiiTax: 0,
    previousDarf: 0,
    payableTax: 0,
  },
};

export const getMonthNumber = (monthName: string) : number => {
  const index = allTabs.findIndex((element) => element === monthName);
  return index === -1 ? 0 : (index);
};

export function getTaxDetailsInfo(monthlyTaxReport ?: MonthlyTaxReport) : ITaxdDetailInfo {
  if (monthlyTaxReport) {
    const commonOpsValidate = monthlyTaxReport.commonOpsReport
      && ((monthlyTaxReport.commonOpsReport.totalSaleVolume !== 0)
      || (monthlyTaxReport.commonOpsReport.profits !== 0)
      || (monthlyTaxReport.commonOpsReport.taxableProfits !== 0)
      || (monthlyTaxReport.commonOpsReport.previousMonthLosses !== 0)
      || (monthlyTaxReport.commonOpsReport.taxBase !== 0)
      || (monthlyTaxReport.commonOpsReport.taxOwed !== 0));

    const commonOps = monthlyTaxReport.commonOpsReport && commonOpsValidate ? {
      totalSell: monthlyTaxReport.commonOpsReport.totalSaleVolume,
      monthProfit: monthlyTaxReport.commonOpsReport.profits,
      taxableProfit: monthlyTaxReport.commonOpsReport.taxableProfits,
      compensatedLoss: monthlyTaxReport.commonOpsReport.previousMonthLosses,
      calcBase: monthlyTaxReport.commonOpsReport.taxBase,
      taxDue: monthlyTaxReport.commonOpsReport.taxOwed,
    } : undefined;

    const dayTradeValidate = monthlyTaxReport.dayTradeReport
      && ((monthlyTaxReport.dayTradeReport.profits !== 0)
      || (monthlyTaxReport.dayTradeReport.previousMonthLosses !== 0)
      || (monthlyTaxReport.dayTradeReport.taxBase !== 0)
      || (monthlyTaxReport.dayTradeReport.taxesOwed !== 0));

    const dayTrade = monthlyTaxReport.dayTradeReport && dayTradeValidate ? {
      taxableProfit: monthlyTaxReport.dayTradeReport.profits,
      compensatedLoss: monthlyTaxReport.dayTradeReport.previousMonthLosses,
      calcBase: monthlyTaxReport.dayTradeReport.taxBase,
      taxDue: monthlyTaxReport.dayTradeReport.taxesOwed,
    } : undefined;

    const fiiValidate = monthlyTaxReport.fiiReport
      && ((monthlyTaxReport.fiiReport.profits !== 0)
      || (monthlyTaxReport.fiiReport.previousMonthLosses !== 0)
      || (monthlyTaxReport.fiiReport.taxBase !== 0)
      || (monthlyTaxReport.fiiReport.netTaxes !== 0));

    const fii = monthlyTaxReport.fiiReport && fiiValidate ? {
      liquidResult: monthlyTaxReport.fiiReport.profits,
      compensatedLoss: monthlyTaxReport.fiiReport.previousMonthLosses,
      calcBase: monthlyTaxReport.fiiReport.taxBase,
      taxDue: monthlyTaxReport.fiiReport.netTaxes,
    } : undefined;

    const finalResults = {
      commonOpTax: monthlyTaxReport.commonOpsReport?.taxOwed || 0,
      daytradeTax: monthlyTaxReport.dayTradeReport?.taxesOwed || 0,

      withholdingTax: (monthlyTaxReport.commonOpsReport?.irrfToCompensate || 0)
        + (monthlyTaxReport.dayTradeReport?.irrfToCompensate || 0),

      totalTax: monthlyTaxReport.stockMarketNetTaxes,
      fiiTax: monthlyTaxReport.fiiReport?.netTaxes || 0,
      previousDarf: monthlyTaxReport.previousDarf,

      payableTax: (monthlyTaxReport.fiiReport?.netTaxes || 0)
        + monthlyTaxReport.stockMarketNetTaxes
        + monthlyTaxReport.previousDarf,
    };

    return {
      commonOps,
      dayTrade,
      fii,
      finalResults,
    };
  }

  return defaultTaxdDetailInfo;
}

/**
 * Search at TaxReport redux state if contains the report of this year of input monthNumber
 * @param taxReport: Tax Report Redux State from API
 * @param monthNumber: Month number to get the report (January:1, Feb: 2,..., December: 12)
 */
export const getMonthReport = (monthNumber: number, taxReport?: ITaxReport)
    : MonthlyTaxReport | undefined => {
  const thisYear = new Date().getFullYear();
  if (taxReport) {
    return taxReport.monthlyReports.find((item) => (item.monthIndex === monthNumber - 1)
        && item.year === thisYear);
  }
  return undefined;
};
