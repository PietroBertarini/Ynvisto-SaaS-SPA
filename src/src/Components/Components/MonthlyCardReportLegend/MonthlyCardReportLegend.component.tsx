import React from 'react';
import { Typography } from '@material-ui/core';
import Box from '@material-ui/core/Box';
import Grid from '@material-ui/core/Grid';
import { useStyles } from './MonthlyCardReportLegend.styles';
import {
  sumAllDataWithSameSignAsHeader, percentValue, getTotalValue,
} from './MonthlyCardReportLegend.utils';
import { IMonthlyCardData } from './MonthlyCardReportLegend.types';

export interface OwnProps {
    /**
     * Array of objects that contain : Label, some number value and color for legend.
     */
    legendData : IMonthlyCardData[];

}

export interface FormRowProps {
    color : string;
    key: string;
    label: string;
    value: number;
    denominator: number;
}

/**
 * Generates a color map for each ValuesBar in Monthly Tax card.
 */
function MonthlyCardReportLegend(props: OwnProps) {
  const { legendData } = props;
  const classes = useStyles();
  const totalVolume = getTotalValue(legendData);
  const denominator = sumAllDataWithSameSignAsHeader(legendData, totalVolume);

  return (
    <div className={classes.root}>
      <Box display="flex" flexDirection="column" justifyContent="space-between">
        {legendData.map((item : IMonthlyCardData) => (
          <FormRow
            key={item.label}
            color={item.color}
            label={item.label}
            value={item.value}
            denominator={denominator}
          />
        ))}
      </Box>
    </div>
  );
}

function FormRow(props: FormRowProps) {
  const {
    label, color, value, denominator,
  } = props;

  const classes = useStyles(props);

  return (
    <Grid container item xs={12} spacing={2} alignItems="center" className={classes.row}>
      <Grid item xs={1}>
        <Box borderRadius="50%" bgcolor={color} style={{ width: '1.2vw', height: '1.2vw' }} />
      </Grid>
      <Grid item xs={4}>
        <Typography variant="subtitle1" className={classes.helper}>
          {label}
        </Typography>
      </Grid>
      <Grid item xs={3}>
        <Typography variant="subtitle1" className={classes.helper}>
          {percentValue(value, denominator)}
        </Typography>
      </Grid>
      <Grid item xs={4}>
        <Typography variant="subtitle1" className={classes.helper}>
          {value.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}
        </Typography>
      </Grid>
    </Grid>
  );
}
export default MonthlyCardReportLegend;
