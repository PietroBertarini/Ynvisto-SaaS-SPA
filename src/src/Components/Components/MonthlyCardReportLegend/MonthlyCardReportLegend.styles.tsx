import { createStyles, makeStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(() => createStyles({
  root: {
    paddingBottom: '1.2vw',
  },
  row: {
    marginBottom: '0.8vw',
  },
  shape: {
    width: '1.2vw',
    height: '1.2vw',
    marginRight: '0.6vw',
  },
  shapeCircle: {
    borderRadius: '50%',
  },
  helper: {
    color: '#FFFFFF',
  },
}));
