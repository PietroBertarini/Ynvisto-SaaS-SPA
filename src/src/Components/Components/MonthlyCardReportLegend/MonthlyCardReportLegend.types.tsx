export interface IMonthlyCardData {
  /**
   * Choose a color of single helper.
   */
  color : string;
  /**
   * The label of the helper
   */
  label : string;
  /**
   * The number of the helper
   */
  value : number;
}
