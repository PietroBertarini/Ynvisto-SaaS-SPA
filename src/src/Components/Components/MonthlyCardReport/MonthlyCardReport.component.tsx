import React, { useCallback, useState } from 'react';
import Typography from '@material-ui/core/Typography';
import {
  Card, CardActionArea,
} from '@material-ui/core';
import Box from '@material-ui/core/Box';
import Dialog from '@material-ui/core/Dialog';
import CardContent from '@material-ui/core/CardContent';
import { useStyles } from './MonthlyCardReport.styles';
import ValuesBar from '../ValuesBar/ValuesBar.component';
import MonthlyCardReportLegend from '../MonthlyCardReportLegend/MonthlyCardReportLegend.component';
import { IMonthlyCardData } from '../MonthlyCardReportLegend/MonthlyCardReportLegend.types';

export enum ETitleStatus {
  GOOD = 'GOOD',
  WARNING = 'WARNING',
  BAD = 'BAD',
}

export interface OwnProps {
  /**
   * Header Status. Directly controls the color of the header monetary value
   */
  status: ETitleStatus,
  /**
   * Title of the card
   */
  title: string,
  /**
   * Main monetary value displayed on at the top of the card
   */
  headerMonetaryValue: number,
  /**
   * The data info to construct the legend and the bar of the card.
   */
  data: IMonthlyCardData[],
  /**
   * Modal that will be shown if clicked.
   */
  detailsModal?: JSX.Element,
}

/**
 * Get the stores states that TaxComponents uses and observe.
 * @param state: the stored stated of the Ynvisto.
 */

/**
 * Card with % bar and legend to show some inf.
 */
function MonthlyCardReport(props : OwnProps) {
  const {
    title, headerMonetaryValue, data, detailsModal,
  } = props;
  const [open, setOpen] = useState(false);

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  // Empty function, according by ESlint documentation, to avoid empty arrow functions
  const emptyFunc = useCallback(() => {
    // do nothing
  }, []);

  const classes = useStyles(props);
  return (
    <>
      <Card className={classes.card} onClick={detailsModal ? handleOpen : emptyFunc}>
        <CardActionArea className={classes.actionArea}>
          <CardContent className={classes.titleContainer}>
            <Box
              display="flex"
              justifyContent="center"
              alignContent="center"
              alignItems="center"
              className={classes.titleContainer}
            >
              <Typography variant="h5" align="center" className={classes.title}>
                {title}
              </Typography>
            </Box>
          </CardContent>
          <CardContent className={classes.content}>
            <Box display="flex" flexDirection="column" alignSelf="center" justifyContent="center">
              <Typography variant="h4" className={classes.headerMonetaryValue}>
                {headerMonetaryValue.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}
              </Typography>
              <ValuesBar
                data={data}
              />
              <MonthlyCardReportLegend legendData={data} />
            </Box>
            {detailsModal && (<div className={classes.showDiv}><p className={classes.showMore}>Mais Detalhes</p></div>)}
          </CardContent>
        </CardActionArea>
      </Card>

      <Dialog
        className={classes.modal}
        onClose={handleClose}
        scroll="body"
        PaperProps={{
          style: { borderRadius: 20, backgroundColor: 'transparent' },
        }}
        aria-labelledby="simple-dialog-title"
        open={open}
      >
        <div className={classes.paper}>
          {detailsModal}
        </div>
      </Dialog>
    </>
  );
}

export default MonthlyCardReport;
