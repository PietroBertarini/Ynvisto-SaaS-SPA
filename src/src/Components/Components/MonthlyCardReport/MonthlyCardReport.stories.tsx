import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta, Story } from '@storybook/react/types-6-0';

import MonthlyCardReport, { ETitleStatus, OwnProps } from './MonthlyCardReport.component';
import { getProfitDataDefault } from '../../../Mocks/MonthlyCardReport.mock';

export default {
  title: 'Components/MonthlyCardReport',
  component: MonthlyCardReport,
} as Meta;

const Template: Story<OwnProps> = (args) => <MonthlyCardReport {...args} />;

export const Default = Template.bind({});
Default.args = {
  status: ETitleStatus.GOOD,
  title: 'stories',
  headerMonetaryValue: 10000,
  data: getProfitDataDefault,

};
