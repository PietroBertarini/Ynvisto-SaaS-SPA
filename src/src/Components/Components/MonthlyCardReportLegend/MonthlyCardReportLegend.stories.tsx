import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta, Story } from '@storybook/react/types-6-0';
import MonthlyCardReportLegend, { OwnProps } from './MonthlyCardReportLegend.component';
import { getProfitDataDefault } from '../../../Mocks/MonthlyCardReport.mock';

export default {
  title: 'Components/MonthlyCardReportLegend',
  component: MonthlyCardReportLegend,
} as Meta;

const Template: Story<OwnProps> = (args) => <MonthlyCardReportLegend {...args} />;

export const StatusOk = Template.bind({});
StatusOk.args = {
  legendData: getProfitDataDefault,
};
