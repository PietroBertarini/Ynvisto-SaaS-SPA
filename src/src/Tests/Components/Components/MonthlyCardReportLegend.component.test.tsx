import React from 'react';
import { shallow } from 'enzyme';
import { getProfitData2Vale } from '../../../Mocks/MonthlyCardReport.mock';
import MonthlyCardReportLegend from
  '../../../Components/Components/MonthlyCardReportLegend/MonthlyCardReportLegend.component';

test('Renders MonthlyCardReportLegend  component and match to previously snapshot', () => {
  const component = shallow(<MonthlyCardReportLegend legendData={getProfitData2Vale} />);
  expect(component).toMatchSnapshot();
});
