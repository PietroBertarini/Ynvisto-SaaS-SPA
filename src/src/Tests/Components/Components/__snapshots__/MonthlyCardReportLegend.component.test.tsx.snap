// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders MonthlyCardReportLegend  component and match to previously snapshot 1`] = `
<div
  className="makeStyles-root-1"
>
  <Styled(MuiBox)
    display="flex"
    flexDirection="column"
    justifyContent="space-between"
  >
    <FormRow
      color="#52A9F4"
      denominator={110}
      key="Operações Comuns"
      label="Operações Comuns"
      value={10}
    />
    <FormRow
      color="#71D3DE"
      denominator={110}
      key="Day Trade"
      label="Day Trade"
      value={100}
    />
  </Styled(MuiBox)>
</div>
`;
